src/components/FullCourt/index.tsx

```diff
@@ -38,6 +43,10 @@ const FullCourt = () => {
     ref.current.y(0);
   }, [canvasStates.resetState]);
 
+  useEffect(() => {
+    updateCourtStage(dispatch, ref, rulerState, toast);
+  }, [canvasStates.selectedColor, borderLength]);
+

```

Comment:
Duplicate codes line 17,18,21,22,24,46-49 in all  courts components. Consider using a custom hook and wrapping all of your duplicated codes in it. An example can be found inside of hooks folder, such as the useCourt file.

Reply:
solved! it's useImageDataUrl custom hook now